src/app.rs

```diff
@@ -1,39 +1,71 @@
-use crate::app_builder::AppBuilder;
+use std::any::{type_name, TypeId};
+
+use crate::{
+    app_builder::AppBuilder, type_names::TypeNames, AppWorkload, AppWorkloadInfo, Plugin,
+    TypeIdBuckets,
+};
 use shipyard::*;
-use tracing::*;
+use tracing::trace_span;
 
 #[allow(clippy::needless_doctest_main)]
 /// Containers of app logic and data
 pub struct App {
     pub world: World,
-    pub(crate) update_stage: &'static str,
-    // pub(crate) startup_stages: Vec<&'static str>,
-    // pub resources: Resources,
-    // pub runner: Box<dyn Fn(App)>,
-    // pub schedule: Schedule,
-    // pub executor: ParallelExecutor,
-    // pub startup_schedule: Schedule,
-    // pub startup_executor: ParallelExecutor,
+    pub(crate) type_names: TypeNames,
+    workload_ids: TypeIdBuckets<()>,
 }
 
-// fn run_once(mut app: App) {
-//     app.update();
-// }
-
 impl App {
-    pub fn build() -> AppBuilder {
-        AppBuilder::new()
+    #[allow(clippy::new_without_default)]
+    pub fn new() -> App {
+        App::new_with_world(World::new())
+    }
+    pub fn new_with_world(world: World) -> App {
+        let type_names = TypeNames::default();
+        App {
+            world,
+            workload_ids: TypeIdBuckets::new("Count times workload plugin added", &type_names),
+            type_names,
+        }
     }
 
+    #[track_caller]
+    pub fn add_plugin_workload<P>(&mut self, plugin: P) -> AppWorkload
+    where
+        P: Plugin + 'static,
```

Comment:
I usually put the generics right in the definition if the line doesn't get too wide:

```suggestion
    pub fn add_plugin_workload<P: Plugin + 'static>(&mut self, plugin: P) -> AppWorkload
```

Reply:
The following fn (`add_plugin_workload_with_info`) has the same structure but does become too wide for one line, so I'm leaning towards the two signatures reading the same way.